tools/loader/ImageClassifier.cpp

```diff
@@ -364,6 +372,76 @@ static void parseInputImageList(const std::string &inputImageListFile) {
   inFile.close();
 }
 
+/// Run inference request on HostManager.
+static void runInference(runtime::HostManager *hostManager, std::string name,
+                         std::unique_ptr<ExecutionContext> batch,
+                         std::promise<void> &runPromise,
+                         std::atomic<unsigned> &inflight,
+                         std::atomic<int> &dispatched) {
+  hostManager->runNetwork(name, std::move(batch),
+                          [&runPromise, &inflight, &dispatched, hostManager,
+                           name](runtime::RunIdentifierTy, llvm::Error err,
+                                 std::unique_ptr<ExecutionContext> contextPtr) {
+                            EXIT_ON_ERR(std::move(err));
+                            // Kick off another run.
+                            if (dispatched.fetch_sub(1) > 0) {
+                              inflight++;
+                              runInference(hostManager, name,
+                                           std::move(contextPtr), runPromise,
+                                           inflight, dispatched);
+                            }
+                            if (--inflight == 0) {
+                              runPromise.set_value();
+                            }
+                          });
+}
+
+/// Run benchmark and wait for all runs to complete.
+static void
+runBenchmark(std::string name, Loader &loader,
+             std::vector<std::unique_ptr<ExecutionContext>> contexts,
+             unsigned requestCount) {
+  runtime::HostManager *hostManager = loader.getHostManager();
+  std::atomic<unsigned> inflight(0);
+  std::atomic<int> dispatched(requestCount);
+  std::promise<void> runPromise;
+  auto fut = runPromise.get_future();
+
+  // Kick off initial pool of requests.
+  for (unsigned i = 0; i < contexts.size(); i++) {
+    auto batch = std::move(contexts[i]);
+    inflight++;
+    dispatched--;
+    runInference(hostManager, name, std::move(batch), runPromise, inflight,
+                 dispatched);
+  }
+  // wait for all to finish.
+  fut.wait();
+}
+
+static void
+setupContextPool(Placeholder *outputPH, Placeholder *inputImagePH,
+                 glow::Tensor &inputImageData,
+                 std::vector<std::unique_ptr<Tensor>> &preLoadedImages,
+                 std::vector<std::unique_ptr<ExecutionContext>> &contexts) {
+  // Size of the pool, the smaller of poolSize or the actual number of
+  // requests.
+  unsigned iterations =
+      miniBatch ? std::min(int(poolSize), int(iterationsOpt / miniBatch)) : 1;
+  // Setup pool of inference requests to be run.
+  for (unsigned i = 0; i < iterations; i++) {
+    auto newImage = llvm::make_unique<Tensor>(inputImageData.getType());
+    auto newContext = llvm::make_unique<ExecutionContext>();
+    auto ph = newContext->getPlaceholderBindings();
+    ph->allocate(inputImagePH);
```

Comment:
Yeah so referring back to my other comment, I think instead of allocating here and then calling `updateInputPlaceholders()` you can insert use `PlaceholderBindings::insert()` to insert the `newImage` alongside `inputImagePH`. Then `preLoadedImages` can be removed. Right?